code/modules/paperwork/handlabeler.dm

```diff
@@ -128,3 +155,177 @@
+	applying_to.AddComponent( \
+		/datum/component/sticker, \
+		stickering_atom = src, \
+		dir = applying_to.dir, \
+		px = stick_px, \
+		py = stick_py, \
+		stick_callback = CALLBACK(src, PROC_REF(on_stick)), \
+		peel_callback = CALLBACK(src, PROC_REF(on_peel)), \
+	)
+
+/// Callback invoked when the label is attached to something
+/obj/item/label/proc/on_stick(atom/applying_to)
+	sticking_to = applying_to
+	RegisterSignal(sticking_to, COMSIG_ATOM_ITEM_INTERACTION, PROC_REF(interacted_with))
+	RegisterSignal(sticking_to, COMSIG_ATOM_EXAMINE, PROC_REF(on_examine))
+	RegisterSignals(sticking_to, list(SIGNAL_ADDTRAIT(TRAIT_WAS_RENAMED), SIGNAL_REMOVETRAIT(TRAIT_WAS_RENAMED)), PROC_REF(reapply))
+	RegisterSignal(sticking_to, COMSIG_QDELETING, PROC_REF(clear_stick_to))
+	apply_label()
+
+/// Callback invoked when the label is removed from something
+/obj/item/label/proc/on_peel(atom/peeled_from)
+	qdel(src)
+
+/// General purpose / signal proc used to clear references and clean up when removed
+/obj/item/label/proc/clear_stick_to(...)
+	SIGNAL_HANDLER
+
+	if(isnull(sticking_to))
+		return
+	if(!QDELING(sticking_to))
+		remove_label()
+	UnregisterSignal(sticking_to, list(
+		COMSIG_ATOM_ITEM_INTERACTION,
+		COMSIG_ATOM_EXAMINE,
+		SIGNAL_ADDTRAIT(TRAIT_WAS_RENAMED),
+		SIGNAL_REMOVETRAIT(TRAIT_WAS_RENAMED),
+		COMSIG_QDELETING,
+	))
+	sticking_to = null
+
+/**
+ * This proc will trigger when any object is used to attack the thing we're stuck to. .
+ *
+ * If the attacking object is not a hand labeler, it will return.
+ * If the attacking object is a hand labeler, it will either update the label or remove the label entirely.
+ *
+ * Arguments:
+ * * source: The parent.
+ * * attacker: The object that is hitting the parent.
+ * * user: The mob who is wielding the attacking object.
+*/
+/obj/item/label/proc/interacted_with(atom/source, mob/living/user, obj/item/tool)
+	SIGNAL_HANDLER
+
+	// If the attacking object is not a hand labeler or its mode is 1 (has a label ready to apply), return.
+	// The hand labeler should be off (mode is 0), in order to remove a label.
+	var/obj/item/hand_labeler/labeler = tool
+	if(!istype(labeler))
+		return NONE
+
+	if(labeler.mode)
+		if(!length(labeler.label))
+			labeler.balloon_alert(user, "no text set!")
+			return ITEM_INTERACT_BLOCKING
+		if(labeler.label == label_name)
+			sticking_to.balloon_alert(user, "already labelled!")
+			return ITEM_INTERACT_BLOCKING
+		if(length(initial(sticking_to.name)) + length(labeler.label) > MAX_LABEL_LEN)
+			sticking_to.balloon_alert(user, "label too long!")
+			return ITEM_INTERACT_BLOCKING
+
+		update_label_name(labeler.label)
+		playsound(sticking_to, 'sound/items/handling/component_pickup.ogg', 20, TRUE)
+		sticking_to.balloon_alert(user, "label renamed")
+	else
+		playsound(sticking_to, 'sound/items/poster_ripped.ogg', 20, TRUE)
+		sticking_to.balloon_alert(user, "label removed")
+		qdel(src)
+	return ITEM_INTERACT_SUCCESS
```

Comment:
i think it would be nice if this section was an early return instead because i feel the stuff in the `if` condition is needlessly indented. your call though